File: java/azure-storage-blob-perfstress/src/main/java/com/azure/storage/blob/perfstress/UploadFromFileTest.java
package com.azure.storage.blob.perfstress;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;

import com.azure.perfstress.RandomStream;
import com.azure.perfstress.SizeOptions;
import com.azure.storage.blob.perfstress.core.RandomBlobTest;

import reactor.core.publisher.Mono;

public class UploadFromFileTest extends RandomBlobTest<SizeOptions> {

    private static Path tempFile;

    public UploadFromFileTest(SizeOptions options) {
        super(options);
    }

    @Override
    public Mono<Void> GlobalSetupAsync() {
        return super.GlobalSetupAsync().then(CreateTempFile());
    }

    @Override
    public Mono<Void> GlobalCleanupAsync() {
        return DeleteTempFile().then(super.GlobalCleanupAsync());
    }

    private Mono<Void> CreateTempFile() {
        try {
            tempFile = Files.createTempFile(null, null);
            
            InputStream inputStream = RandomStream.create(Options.Size);
            OutputStream outputStream = new FileOutputStream(tempFile.toString());
            inputStream.transferTo(outputStream);
            outputStream.close();
            
            return Mono.empty();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private Mono<Void> DeleteTempFile() {
        try {
            Files.delete(tempFile);
            return Mono.empty();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void Run() {
        try {
            _blobClient.uploadFromFile(tempFile.toString());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Mono<Void> RunAsync() {
        return _blobAsyncClient.uploadFromFile(tempFile.toString());
    }
}

File: net/Azure.Storage.Blobs.PerfStress/UploadFromFileTest.cs
using Azure.Storage.Blobs.PerfStress.Core;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace Azure.Storage.Blobs.PerfStress
{
    public class UploadFromFileTest : RandomBlobTest<StorageTransferOptionsOptions>
    {
        private static string _tempFile;

        public UploadFromFileTest(StorageTransferOptionsOptions options) : base(options)
        {
        }

        public override async Task GlobalSetupAsync()
        {
            await base.GlobalSetupAsync();

            _tempFile = Path.GetTempFileName();

            using var randomStream = RandomStream.Create(Options.Size);
            using var fileStream = File.OpenWrite(_tempFile);
            await randomStream.CopyToAsync(fileStream);
            fileStream.Close();
        }

        public override async Task GlobalCleanupAsync()
        {
            File.Delete(_tempFile);
            await base.GlobalCleanupAsync();
        }

        public override void Run(CancellationToken cancellationToken)
        {
            BlobClient.Upload(_tempFile, transferOptions: Options.StorageTransferOptions, cancellationToken: cancellationToken);
        }

        public override Task RunAsync(CancellationToken cancellationToken)
        {
            return BlobClient.UploadAsync(_tempFile, transferOptions: Options.StorageTransferOptions, cancellationToken: cancellationToken);
        }
    }
}
